planning/behavior_path_planner/src/utils/drivable_area_expansion/drivable_area_expansion.cpp

```diff
@@ -129,8 +129,32 @@ std::array<ring_t::const_iterator, 4> findLeftRightRanges(
   return {left_start, left_end, right_start, right_end};
 }
 
+void copy_z_over_arc_length(
+  const std::vector<geometry_msgs::msg::Point> & from, std::vector<geometry_msgs::msg::Point> & to)
+{
+  if (from.empty() || to.empty()) return;
+  to.front().z = from.front().z;
+  if (from.size() < 2 || to.size() < 2) return;
+  to.back().z = from.back().z;
+  auto i_from = 1lu;
+  auto s_from = tier4_autoware_utils::calcDistance2d(from[0], from[1]);
+  auto s_to = 0.0;
+  auto s_from_prev = 0.0;
+  for (auto i_to = 1lu; i_to + 1 < to.size(); ++i_to) {
+    s_to += tier4_autoware_utils::calcDistance2d(to[i_to - 1], to[i_to]);
+    for (; s_from < s_to && i_from + 1 < from.size(); ++i_from) {
+      s_from_prev = s_from;
+      s_from += tier4_autoware_utils::calcDistance2d(from[i_from], from[i_from + 1]);
+    }
+    const auto ratio = (s_to - s_from_prev) / (s_from - s_from_prev);
+    to[i_to].z = from[i_from - 1].z + ratio * (from[i_from].z - from[i_from - 1].z);
```

Comment:
You can use linear interpolation library from interpolation package
https://github.com/autowarefoundation/autoware.universe/tree/main/common/interpolation

Also, I guess you have to add a guard when `s_from=s_from_prev`

Reply:
Done: [31d72bd](https://github.com/autowarefoundation/autoware.universe/pull/4120/commits/31d72bdc06c99cd0799fefdefd7c0fbfa0300db3)